Storage/SQL/SQLiteBookmarksSyncing.swift

```diff
@@ -1189,4 +1224,144 @@ extension MergedSQLiteBookmarks {
+        let resultError = self.local.db.transaction(&err) { (conn, err: inout NSError?) in
+            func change(_ sql: String, args: Args?=nil) -> Bool {
+                if let e = conn.executeChange(sql, withArgs: args) {
+                    err = e
+                    deferred.fillIfUnfilled(Maybe(failure: DatabaseError(err: e)))
+                    return false
+                }
+                return true
+            }
+
+            // So we can trample the DB in any order.
+            if !change("PRAGMA defer_foreign_keys = ON") {
+                return false
+            }
+
+            guard let mobileRoot = op.mobileRootToUpdate else {
+                return false
+            }
+
+            let insertSQL = [
+                "INSERT OR REPLACE INTO \(TableBookmarksBuffer) (",
+                "type, server_modified, is_deleted,",
+                "hasDupe, parentid, parentName,",
+                "feedUri, siteUri, pos, title,",
+                "description, bmkUri, tags, keyword,",
+                "folderName, queryId, guid)",
+                "VALUES",
+                "(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
+                ].joined(separator: " ")
+
+            let args = mobileRoot.getUpdateOrInsertArgs()
+            _ = change(insertSQL, args: args)
+
+            if err != nil {
+                return false
+            }
+
+            op.bufferValuesToMoveFromLocal
+                .withSubsetsOfSize(BrowserDB.MaxVariableNumber) { guids in
+                    let args: Args = guids.map { $0 }
+                    let varlist = BrowserDB.varlist(guids.count)
+
+                    let copySQL = [
+                        "INSERT OR REPLACE INTO \(TableBookmarksBuffer)",
+                        "(guid, type, parentid, parentName, feedUri, siteUri, pos, title, description,",
+                        "bmkUri, tags, keyword, folderName, queryId, server_modified)",
+                        "SELECT guid, type, parentid, parentName, feedUri, siteUri, pos, title, description,",
+                        "bmkUri, tags, keyword, folderName, queryId,",
+
+                        // This will be fixed up in batches after the initial copy.
+                        "0 AS server_modified",
+                        "FROM \(TableBookmarksLocal) WHERE guid IN",
+                        varlist
+                        ].joined(separator: " ")
+                    _ = change(copySQL, args: args)
+            }
+
+            if err != nil {
+                return false
+            }
+
+            op.bufferValuesToMoveFromLocal
+                .withSubsetsOfSize(BrowserDB.MaxVariableNumber) { guids in
+                    guard err == nil else { return }
+                    let args: Args = guids.map { $0 }
+                    let varlist = BrowserDB.varlist(guids.count)
+
+                    let sqlLocalStructure = "DELETE FROM \(TableBookmarksLocalStructure) WHERE parent IN \(varlist)"
```

Comment:
I think in this case that should be `child`, not `parent`, for two reasons:

- You're not uploading folders.
- We've just moved the record to the buffer, and we're forgetting the local copy. You need this to be `child` to unlink it.

N.B., with this cleanup it means skipped can leave `idx` sparse.